### ISSUE TITLE
make option to exclude packages

### ISSUE THREAD

**Issue by zteitler** [Macaulay2/M2#355]
Minor feature request: I have been having trouble with RunExternalM2 (fails `make check` on my laptop). It would be nice to have an option like `EXCLUDEPACKAGES='Foo Bar'` to exclude Foo and Bar from the list of packages to build, check, and install. For example this might be possible via gmake's filter-out command <https://www.gnu.org/software/make/manual/html_node/Text-Functions.html>; something like `PACKAGES = $(filter-out $(EXCLUDEPACKAGES),$(PACKAGES))`. I tried to get this working but I could not figure out what file to put it in (configure.ac? Macaulay2/packages/Makefile.in?). A different or shorter name than EXCLUDEPACKAGES would be fine.

_edit: I shouldn't have said RunExternalM2 was 'broken'._

**Comment by DanGrayson** [comment]
"make" already has an option for skipping a target:"-o".  If "-o
check-RunExternalM2" were
on the command line when the makefile in M2/Macaulay2 runs the makefile in
M2/Macaulay2/packages, then that might work.  That would be much cleaner
that futzing
with filtering strings.  Try something like that.

A long term goal of ours is to have just one master makefile, so we don't
call "make" recursively,
except for libraries.  If that had already been done, there would be no
issue with propagating
command line options.

**Comment by DanGrayson** [comment]
Bypassing checks is not so great, anyway, as errors should be fixed.
RunExternalM2
passes its tests at my end.  What OS are you on?

**Comment by zteitler** [comment]
I'm on El Capitan, 10.11.1.

$ gcc-4.9 --version
gcc-4.9 (Homebrew gcc49 4.9.3) 4.9.3

**Comment by DanGrayson** [comment]
Mike, have you already upgraded to El Capitan?

**Comment by mikestillman** [comment]
No, I have not.  Have you?

**Comment by DanGrayson** [comment]
Nope.  Are you willing?

**Comment by zteitler** [comment]
For what it's worth, `make -C Macaulay2/packages/ check -o check-RunExternalM2` worked. Thanks Dan. (`make check -o check-RunExternalM2` didn't seem to work, or perhaps I did something wrong.)

**Comment by DanGrayson** [comment]
Yes, I said that other one wouldn't work ...

**Comment by bapike** [comment]
I'd like to fix `RunExternalM2`, if possible.  Can you tell me which test(s) fail and provide the .tmp file(s) that gives the output?

If one of the broken tests is test 6, specifically the line where `KeepStatistics=>true`, then please also send the output of this script:
```
ls -al /usr/bin/time
/usr/bin/time --version
echo $?
ls -al /usr/bin/sw_vers
/usr/bin/sw_vers -productName
echo $?
(time sh -c "")
echo $?
```

Thanks!

**Comment by DanGrayson** [comment]
test 1 fails, and I attach the file

**Comment by bapike** [comment]
Sorry, no file made it.  Apparently you can attach a file to a github issue or pull request by dragging and dropping it into the comment box, but maybe not e-mail?

**Comment by DanGrayson** [comment]
Oh, I can't just expect attachments on email messages to survive, I guess.  Here it is again:

[16.txt](https://github.com/Macaulay2/M2/files/65695/16.txt)

**Comment by bapike** [comment]
I have a feeling that the same behavior would be observed by something like:
```
useSomeMemory = (n) -> (
  R:=QQ[x,y,z,w];
  M:=random(R^n,R^{n:-2});
);

testFork = (n) -> (
  pid:=fork();
  if (pid==0) then (
    -- child
    for i from 0 when i<n do (
      useSomeMemory(n);
    );
    collectGarbage();
    exit(0);
  ) else (
    -- parent
    wait(pid);
  )
);

for i from 0 when i<10^3 do (
  testFork(20);
);
```

The "thread_suspend failed" message comes from the Boehm GC, in [bdwgc/darwin_stop_world.c:610](https://github.com/ivmai/bdwgc/blob/51e6d83ec158997ad0e4e813f624f0dee47e33bb/darwin_stop_world.c#L610), when it tries and fails to suspend a thread, probably before trying to collect garbage.

I've only recently realized that M2 actually uses multiple threads internally (or, GDB says so).  My understanding is that after calling `fork()` in a threaded application, only a few specific calls are safe to use in the child: only the thread that calls `fork()` is duplicated, and at that moment other threads could be in an unsafe state, e.g., holding mutexes or other things.  Even `malloc()` should be avoided, apparently, and [M2/Macaulay2/d/stdio.d:256-257](https://github.com/Macaulay2/M2/blob/b54f2c41f9e215e857642e6cfee6ba271fcd7cf0/M2/Macaulay2/d/stdio.d#L256-L257) asserts the same is true of `GC_malloc`.

Nevertheless, [bdwgc/include/gc.h](https://github.com/ivmai/bdwgc/blob/51e6d83ec158997ad0e4e813f624f0dee47e33bb/include/gc.h#L427-L436) has routines `GC_atfork_{prepare,parent,child}` that should be called before and just after forking, to leave the GC in a safe state, and these routines are not explicitly used in Macaulay2.  However, according to [the comments before `GC_set_handle_fork` in bdwgc/include/gc.h](https://github.com/ivmai/bdwgc/blob/51e6d83ec158997ad0e4e813f624f0dee47e33bb/include/gc.h#L413-L425), the GC may attempt to use `pthread_atfork` to force the calling of these routines, and maybe this is successful on most platforms?

My guess is that the `pthread_atfork` is failing on the newer OS X, hence the GC in the child has a list of threads that don't actually exist, can't be stopped, and so we get the message.

At the least, I think these changes are appropriate:
(1) Everywhere M2 uses `fork()`, it should use `GC_atfork_{prepare,parent,child}` explicitly.
(2) Change RunExternalM2's `runExternalM2InClone` so it checks if `fork()` failed.
(3) Change RunExternalM2 documentation to warn that users shouldn't do much in a child.

But it might be more correct to discourage M2 users from using `fork()`, perhaps by
(4) Removing RunExternalM2's `runExternalM2InClone` function.
(5) Adding documentation to M2's `fork()` to warn users
I don't really want to do (4), since it mostly works.

More generally, what are all the M2 threads doing most of the time (since I don't think there's much threaded code yet), and what should happen to them when using fork()?

Dan, does my shorter program also trigger the problem? Should I send you a commit that does (1)?

**Comment by garyfurnish** [comment]
Is there a reason to prefer fork to starting a thread and running calling system/popen/equivalent?  This is much less platform dependent.  We can't guarantee fork safety in external libs even if it can be forced to work with boehmgc.

**Comment by bapike** [comment]
The goal of the package is to enforce ulimits in an elegant way.  For example, you can force the computation of a randomly-generated example to timeout.  (In my experience, alarm() does not always work.)

The function `runExternalM2` starts a brand-new M2 process with M2's `run()`, executes code in it, lets it terminate, and retrieves the result, much like what you are suggesting.  That's definitely safe.

The function `runExternalM2InClone` forks, runs code in the fork, exits the forked process, and then retrieves the result.  Although using ulimits here is a little awkward (as you must set them before starting the original M2 process), this has the advantage that it is easier to use.  Unfortunately, it may be unsafe because the original M2 has multiple threads.  From what I've read of the code, M2 starts a thread pool where the threads wait for tasks to run.  One such task is the interpreter.  In a "typical for me" situation where no other tasks have been scheduled, there will be:
* the original thread of execution that merely waits for the interpreter to finish
* one thread executing the interpreter; this one will execute the `fork`
* the other threads will be in [ThreadSupervisor::getTask](https://github.com/Macaulay2/M2/blob/b54f2c41f9e215e857642e6cfee6ba271fcd7cf0/M2/Macaulay2/system/supervisor.cpp#L328-L345), probably executing ` pthread_cond_wait`

The last type of thread should not have used any external library functions.  One of them may at the moment of forking be holding a mutex that is used for scheduling, but the interpreter thread should not need this unless the user tries to use M2's threading functions (`schedule`, etc.) in the child M2 process; I already have a warning in the documentation about mixing threading and `runExternalM2InClone`.

Assuming we are able to fix any thread-aware libraries (such as boehmgc), wouldn't this be safe?  It is true that forking will not work on certain platforms, but (2) should fix the code and I should also
(6) fix RunExternalM2's tests to detect platforms where fork does not work

I don't care if this functionality is kept or not.  If we can't get it working on newer OSX, it should be removed.

**Comment by garyfurnish** [comment]
Well one of the bigger concerns is this could silently break if any libraries ever use threadlocal statics or openmp for performance in the future, and this would not be fixable without extensive patching of third party libraries.  Furthermore in general this can only be fixed if the third party library serializes on some global variable which is a major potential performance hit.  No extra thread awareness besides use of a single keyword is necessary to break 2, so for the general case you can't just call something like boehm's on fork functionality.  This had the potential to break even before the original M2 has multiple threads, it is just obvious now.  

I'm not a user of the functionality, but its incredibly dangerous, hard to maintain, and hard to debug, so I tend to be against it if there is not a 100% use case.  (I guess my understanding is we can use ulimits without serialization in one direction?)  I'm not sure I am going to/can safely support it in my drop in replacement for boehm because it is so hard to test and get bug free.  So maybe that is an argument one way or the other.  

"My guess is that the pthread_atfork is failing on the newer OS X, hence the GC in the child has a list of threads that don't actually exist, can't be stopped, and so we get the message."  OSX pthreads and signals and fork do not play nicely together in my experience.  It wouldn't surprise me if there were corner cases either in boehm or OSX kernel/libc++ that are not handled correctly. This is not particularly a problem that is easy to debug speaking from experience.

**Comment by DanGrayson** [comment]
I just tried adding those GC calls for forking, and it didn't help:

diff --git a/M2/Macaulay2/d/actors4.d b/M2/Macaulay2/d/actors4.d
index 6d4d2ce..5edbc41 100644
--- a/M2/Macaulay2/d/actors4.d
+++ b/M2/Macaulay2/d/actors4.d
@@ -30,7 +30,11 @@ forkfun(e:Expr):Expr := (
      is a:Sequence do (
   if length(a) == 0
   then (
+       Ccode(void, "GC_atfork_prepare()");
        pid := fork();
+       if pid == 0
+       then Ccode(void, "GC_atfork_parent()")
+       else Ccode(void, "GC_atfork_child(), GC_start_mark_threads()");
        toExpr(pid))
   else WrongNumArgs(0))
      else WrongNumArgs(0));

Brian, would you be willing to update your package so it doesn't use "fork"?

Enforcing resource limits is better done with "run", anyway, via an
explicit call
to the "ulimit" command of the shell.  If necessary, we can add more
built-in M2 routines
that call setrlimit the way limitFiles() and limitProcesses() do, so we
don't have to
deal with differences between "ulimit" on various systems.

**Comment by bapike** [comment]
I'll update RunExternalM2 to remove the fork functionality.  I'll also [update the documentation for fork](https://github.com/bapike/M2/commit/617d6954c30ffc5eb08c6b10be6d328b86bc91ef) to be more discouraging.  I agree it may be useful to add M2 functions to set limits from within M2.

Dan, does the script I posted here a few weeks ago (containing the string "testFork") also crash?  There are apparently other packages that use fork, like SCSCP; these changes should make the tests pass but may not fix the underlying problem.

Thanks,
Brian

**Comment by DanGrayson** [comment]
Yes, it does fail, repeatedly giving the message "thread_suspend failed"
